feat: consult and edit an administrative unit's scope of operation

mirdono opened the pull request:
OP is extended with functionality to allow users to consult and edit the scope of operation (*nl. werkingsgebied*) for an administrative unit. This PR introduces the necessary frontend changes to use a newly introduced backend service for this functionality.

## Overview of functionality for MVP
### Consult scope of operation
A new field is added to the core data page for organisations to display the organisation's scope of operation. The expected value for this field is a list of all municipality-level locations that are part of the organisation's scope of operation. (Yes this can become very long, just try opening a province :worried:)

### Edit scope of operation
The form to edit the core data for an organisation should allow users to select one or more municipality-level locations as its scope of operation. Similarly, the new organisation form should allow users to select appropriate locations when creating a new administrative unit.

The scope of operation is mandatory for all administrative units and an error should be shown if the users fails to select at least one value.

For municipalities and provinces their scope of operation cannot be edited.

## Proposed solution
This PR introduces a new service in the frontend that acts as a wrapper around the new `scope-of-operation-service` introduced in the backend. This wrapper service essentially translates between the frontend's models and the UUIDs required/returned by the backend service. Putting this logic within a single service allows to reuse it within different routes, such as the edit or new organisation route, without having to duplicate this logic.

Furthermore, a new component is introduced that allows users to select multiple municipality-level locations. This component is similar to the already existing components to select multiple organisations. This new component is added to the edit core data and new organisation form.

## Determine region for organisations
For some kinds organisations (municipalities, IGSs, PEVAs, a OCMW associations) their core data page contains an optional region field. The value of this field is also based on the location that is assigned as the organisation' scope of operation. More specifically, for municipalities it is the reference region in which its corresponding location is located. For the other kinds of organisations it is the reference region of the municipality in which their primary site is located.

The existing functionality for this mapping broke due to a required model change in `LocationModel` to allow that a location is contained within multiple other locations. Consequently, it was no longer possible to simply follow the `locatedWithin` relation as this changed to a `hasMany`. Therefore, this functionality was also revised to search for the reference region in which the municipality is located. This relies on the fact a municipality is at most in 1 reference regions, thus that there is no overlap between reference regions.

## Related PRs:
- Backend: [#550](https://github.com/lblod/app-organization-portal/pull/550)
- Service: [#1](https://github.com/lblod/scope-of-operation-service/pull/1)

## Related tickets:
- OP-3203
- OP-3600

mirdono committed:
feat: extend core data page with a field for scope of operation

Administrative units have a scope of operation (*nl. werkingsgebied*), typically
a set of municipalities. This extends the core data page with an appropriate
field.

mirdono committed:
feat: allow a location to be contained in multiple other locations

The data model specifies that an administrative unit can only have a single
location as its scope of operation. As a consequence new location resources will
have to be created whenever an administrative unit can operate in multiple
municipalities. Therefore, we also need to allow that a single location is
contained within multiple other locations, otherwise it would not be always
possible to create the necessary location resources.

For example, say `UnitOne` has as scope of operation municipalities `A` and
`B`. This requires create a new location resource that contains `A` and
`B`. Furthermore, say `UnitTwo` has as scope of operation municipalities `A` and
`C`. This requires creating a new location resource containing `A` and `C`. If a
location would be allowed to be only contained in one other location that would
not be possible as `A` is already contained in the former location resource.

mirdono committed:
feat: add a component to support selecting multiple locations

This component allows users to select multiple locations as scope of operation
for an organisation. The options in the component are **all** municipality-level
locations, grouped by the provinces they are located in.

As part of the MVP, we do not yet consider district-level locations.

mirdono committed:
feat: allow editing an administrative unit's scope of operation

Extend the form to edit the core data for an organisation with to allow users to
edit the scope of operation for administrative units. Since this is a required
field, the user must select at least one location. otherwise a validation error
is triggered upon saving.

mirdono committed:
feat: specify scope of operation for new administrative units

When creating a new administrative unit, users must specify at least one
location as scope of operation.

mirdono committed:
fix: find appropriate reference region for organizations

The logic to determine the reference region an organization is located in broke
due changing the `locatedWithin` relations from a `belongsTo` to a
`hasMany`. This commit fixes that by searching for a reference region among all
the `locatedWithin` locations instead. This does assume that the reference
regions remain mutually exclusive such that an organisation is located in at
most one reference region.

Furthermore, the template was slightly simplified by removing the duplicate
region entries. The template now relies on a model getter to determine whether
to show the region and always displays the `region` set in the route (instead of
following model relations for municipalities.)

Finally, for municipalities the region is now displayed as a disabled input
field instead of a plain string. This allows to remove one of the region
sections in the template. This is also more consistent with the other fields in
the form.

mirdono committed:
fix: split requests with too long URI into multiple requests

In rare cases a scope of operation contains a large amount of locations, for
example provinces whose scope is all its municipalities. This can result is
calls to the backend that are too long leading to 414 errors from the identifier
or dispatcher service.

Since this is currently a rare occurrence, we opted to manually split the
requests when a large number, more than 20, of locations are requested.

mirdono committed:
feat: add service wrapping backend scope-of-operation-service

An an ember service to wrap the backend's scope-of-operation service. This
frontend service allows to use the functionality of the backend service in
different routes without having to repeat the same boilerplate code, such as
translating the replies from the backend to model elements usable in the frontend.

mirdono committed:
feat: retrieve locations in scope and scope for locations

Extend the scope service to cover the API of the corresponding backend
service. This allows now to
- retrieve all municipality-level locations within an organisation's scope of
operation; and
- retrieve a location resource that contains exactly a set of other location
resources, to that can be used as scope of operation

mirdono committed:
chore: update tests to include scope as mandatory relation

claire-lovisa reviewed (changes requested):
I did find some strange things while testing, putting them below:
- I'm on the edit core data page of Aalst http://localhost:4200/organisaties/974816591f269bb7d74aa1720922651529f3d3b2a787f5c60b73e5a0384950a4/kerngegevens/edit , try to go back to the main page via the breadcrumb "Organisatie", error in the console `route-not-found`. Doesn't seem to happen on the dev branch.
- Still when on the edit page, I can't switch mock login account, the same error gets raised
- Is there a plan to add locations to worship services? The ones I checked didn't have locations and it blocks the update of those worship services, as it's mandatory to be able to save

mirdono commented:
> I did find some strange things while testing, putting them below:
> - I'm on the edit core data page of Aalst http://localhost:4200/organisaties/974816591f269bb7d74aa1720922651529f3d3b2a787f5c60b73e5a0384950a4/kerngegevens/edit , try to go back to the main page via the breadcrumb "Organisatie", error in the console `route-not-found`. Doesn't seem to happen on the dev branch.
> - Still when on the edit page, I can't switch mock login account, the same error gets raised

I am unable to replicate this behaviour in my setup. I did indeed ran into this error occasionally while developing/testing but could never pin down the cause or replicate it reliably. Does the error happen consistently for you in these situations?

> - Is there a plan to add locations to worship services? The ones I checked didn't have locations and it blocks the update of those worship services, as it's mandatory to be able to save

Business initially said it was mandatory for all administrative units. But it seems indeed strange for the case of (central) worship services. I asked for clarification on this in the ticket.

mirdono commented:
Aha, when I open the edit core data page for a province and try to leave by cancelling or via the breadcrumb I do reliably get errors. More specifically: `Error while processing route: route-not-found _ember_data_store_private__WEBPACK_IMPORTED_MODULE_5__.peekCache(...) is null`, and a `TypeError` essentially saying the same thing. Is this the error you also encounter?

claire-lovisa commented:
> Aha, when I open the edit core data page for a province and try to leave by cancelling or via the breadcrumb I do reliably get errors. More specifically: `Error while processing route: route-not-found _ember_data_store_private__WEBPACK_IMPORTED_MODULE_5__.peekCache(...) is null`, and a `TypeError` essentially saying the same thing. Is this the error you also encounter?

Yes it's this exact one!!

mirdono committed:
chore: clarified note

mirdono committed:
fix: remove duplicate controller reset

The edit core data route already contained a call to reset its controller in the
`willTransition` action. The addition of an extra `resetController` function
lead to the controller being reset twice, which is some situations leads to
errors crashing the frontend.

claire-lovisa reviewed (approved):
Good catch :sweat_smile:
